### PR TITLE
Option to keep current steering mode

### PR DESCRIPTION
**Pull request by GeorgeEarslight**
Added option to maintain (not reset) crab steering mode.

### CODE REVIEW

**Comment by GeorgeEarslight** [comment]
Doesn't retain the setting between restarts. 
Kindly point me to where I need to store the setting.

**Comment by Tensuko** [comment]
@GeorgeEarslight 
can you explain us a bit, where the use of that option is for ?
What situation and stuff ?

**Comment by GeorgeEarslight** [comment]
@Tensuko Sorry for not describing.

Some vehicles like the JCB4220, Xerion 5000, etc have the option to change steering mode (CTRL-Y).
When you start the CP driver, it resets it to what it defaults to (because CP calls some Giants code).
I give the option for the player to keep the steering he had already selected, before starting the CP driver.

image

**Comment by Tensuko** [comment]
Hmm I get your point for that, but I wonder if it would make more sence to simply use the actual steering mode and make CP not change it, without need an option for that one.

Or an other way:
Make a Setting to decide the steering by controlling the steering mode with CP (just like we do with straw swath for Combine).
You could say, drive always with XY Steeringmode.
Setting for 'resetting yes/no' I am not that sure if we should do it like that.

On the other Hand, Tools like the Sugarbeet Harvester need to be changed to another steering to have better turn results.

**Comment by GeorgeEarslight** [comment]
My reasoning is that you put the option to activated and then choose from the F1 menu what crab you want.
What you propose is also another way to do it. Did you try it?

**Comment by Tensuko** [comment]
No I did not try it as I had no reason to change the steering so far :)
But I get your reason :)
But I would change the name to 'keep current steering' and also the setting KeepCurrentSteering.

Not sure what @pvaiko would say to it but I had named it like that.

Also Remember to add the Translation into all files pls.

**Comment by GeorgeEarslight** [comment]
Sure, I can change the name to make more sense.
I'll add the other translations too, I forgot :)

**Comment by GeorgeEarslight** [comment]
Borked up the files again. I'll do another commit to fix this.

**Comment by GeorgeEarslight** [comment]
OK, I think I've managed to get all files properly.

File: settings.lua [hunk 1]
```diff
@@ -2494,6 +2494,14 @@ function RidgeMarkersAutomatic:isDisabled()
 	return self.vehicle.cp.driver and not AIDriverUtil.hasAIImplementWithSpecialization(self.vehicle, RidgeMarker)
 end
 
+---@class KeepCurrentSteering : BooleanSetting
+KeepCurrentSteering = CpObject(BooleanSetting)
+function KeepCurrentSteering:init(vehicle)
+	BooleanSetting.init(self, 'KeepCurrentSteering', 'COURSEPLAY_KEEP_CURRENT_STEERING',
+			'COURSEPLAY_YES_NO_KEEP_CURRENT_STEERING', vehicle)
```
Comment (pvaiko) [inline settings.lua:2501]: The second text is for the tooltip in the dialogs, so ideally should be called COURSEPLAY_KEEP_CURRENT_STEERING_TOOLTIP and the texts in the translation file should contain an explanation, for example "Courseplay won't not change the current crab steering setting of the vehicle."

File: settings.lua [hunk 2]
```diff
@@ -2494,6 +2494,14 @@ function RidgeMarkersAutomatic:isDisabled()
 	return self.vehicle.cp.driver and not AIDriverUtil.hasAIImplementWithSpecialization(self.vehicle, RidgeMarker)
 end
 
+---@class KeepCurrentSteering : BooleanSetting
+KeepCurrentSteering = CpObject(BooleanSetting)
+function KeepCurrentSteering:init(vehicle)
+	BooleanSetting.init(self, 'KeepCurrentSteering', 'COURSEPLAY_KEEP_CURRENT_STEERING',
```
Comment (pvaiko) [inline settings.lua:2500]: the second argument is the member variable name in the settings and should be lower camelcase like all other settings (hopefully), so keepCurrentSteering.

File: start_stop.lua [hunk 3]
```diff
@@ -210,7 +210,15 @@ function courseplay:start(self)
 		self.cp.driver = UnloadableFieldworkAIDriver.create(self)
 	end
 	StartStopEvent:sendStartEvent(self)
+  -- Remeber crab steering if relevant
```
Comment (pvaiko) [inline start_stop.lua:213]: So why is this code here, we already have a part dealing with crab steering a few lines above this (181)?

**Review by pvaiko — CHANGES_REQUESTED** [review]

File: settings.lua (hunk 1 above)
Comment (GeorgeEarslight) [reply settings.lua:2501]: Done on next commit.

File: settings.lua (hunk 2 above)
Comment (GeorgeEarslight) [reply settings.lua:2500]: Done on next commit.

File: start_stop.lua (hunk 3 above)
Comment (GeorgeEarslight) [reply start_stop.lua:213]: I've never managed to get through the if clause on L181 even when having a crab capable vehicle.
The debugger didn't help as for some reason it detects 2 self variables and chooses to show only the other.
Finally, even if you did manipulate the mode up there, it's changed later by self.cp.driver:start calling some Giants code for the AI helper. You've looked on the subject some (a lot of) time ago at #3062, not sure if this helps.

Comment (pvaiko) [reply start_stop.lua:213]: Ok, I see, thanks for the link. 

In that case however, the proper solution is to override the CrabSteering:onAIImplementStart() Giants function, like we do with some others in FieldworkAIDriver:register().

Comment (GeorgeEarslight) [reply start_stop.lua:213]: I understand. I'll have to read through the source to understand how to deal with this.
Would this be similar to how you return false here?
https://github.com/Courseplay/courseplay/blob/93ec4a3feda28a8fe542abd35d7c6f6069a8b746/FieldworkAIDriver.lua#L98

Comment (pvaiko) [reply start_stop.lua:213]: yeah, pretty much the same thing, just check the keepCurrentSetting in the new onAIImplementStart() and if it is true, don't call the setCrabSteering(), I believe that's all.

**Pull request merged**